crates/wit-bindgen/src/lib.rs

```diff
@@ -162,18 +162,57 @@ impl Wasmtime {
                 gen.generate_trappable_error_types(TypeOwner::Interface(*id));
                 let iface = &resolve.interfaces[*id];
 
+                let wildcard = iface.functions.values().find(|func| "*" == &func.name);
+
+                if let Some(wildcard) = wildcard {
+                    uwriteln!(gen.src, "pub struct WildcardMatch {{");
+                    uwriteln!(gen.src, "name: String,");
+                    uwriteln!(gen.src, "func: wasmtime::component::TypedFunc<(");
+                    for (_, ty) in wildcard.params.iter() {
+                        gen.print_ty(ty, TypeMode::AllBorrowed("'_"));
+                        gen.push_str(", ");
+                    }
+                    uwriteln!(gen.src, "), (");
+                    for ty in wildcard.results.iter_types() {
+                        gen.print_ty(ty, TypeMode::Owned);
+                        gen.push_str(", ");
+                    }
+                    uwriteln!(gen.src, ")>,");
+                    uwriteln!(gen.src, "}}");
+                    uwriteln!(gen.src, "impl WildcardMatch {{");
+                    gen.define_rust_guest_export(Some(name), wildcard);
+                    uwriteln!(gen.src, "}}");
+                }
+
                 let camel = to_rust_upper_camel_case(name);
                 uwriteln!(gen.src, "pub struct {camel} {{");
                 for (_, func) in iface.functions.iter() {
-                    uwriteln!(
-                        gen.src,
-                        "{}: wasmtime::component::Func,",
-                        func.name.to_snake_case()
-                    );
+                    if "*" == &func.name {
+                        uwriteln!(
+                            gen.src,
+                            "__wildcard_matches: std::collections::HashMap<String, WildcardMatch>,"
```

Comment:
Do y'all have a concrete use case for wildcard exports right now? I ask because if possible I'd like to be conservative and not have them supported. If you need them though, that's fine of course!

The other reason I bring this up is that so far we've avoided baking in hashmap-based string lookups in the component model in Wasmtime and in various support systems. That could sort of be fixed here maybe with something like more generics or similar but it may also be inevitable to some degree. Ideally though the `HashMap` here could be avoided along with the string lookups at runtime for invoking methods.

Reply:
We don't have an immediate need for them yet, no.  I implemented them mainly because they were in scope for https://github.com/WebAssembly/component-model/issues/172.  I also just wanted some experience implementing them to surface any issues that might require design work.

Note that the `__wildcard_matches` variable is private, so we're not committing to using `HashMap` forever.  If we want to get rid of it now, though, we could instead have `instantiate{_pre}` take a `Vec<&str>` containing the names of each export the host wants access to and return a `Vec<WildcardMatch>`.  Would you prefer that?  I'm open to other ideas.

I'd rather not postpone implementing wildcard exports altogether, though.  We've got momentum now, so I'd rather just get it done than come back to it later.

Reply:
Ok keeping the implementation makes sense. I think it's fine to keep this in the back of our heads though and possibly defer to later. The issue isn't so much with the `HashMap` itself but moreso that "some string-related lookup is happening" during instantiation and execution. That's not to say that _zero_ happen today when using Wasmtime though since you're already looking up exports via string lookups, so while it would be nice to avoid I think keeping the simpler interface you've got going here is the way to start.

We can always revisit this later if necessary.

Reply:
Sounds good.  I'll leave it as is for now, but one other thought while we're on the topic: we could change the `HashMap` to a `Vec<WildcardMatch>`, add a `fn name() -> &str` accessor to `WildcardMatch`, and change `fn get_wildcard_match(&self, name: &str) -> Option<&WildcardMatch>` to `get_wildcard_matches(&self) -> &[Wildcard]`.  Then there's no string lookup anywhere, but if the embedder _wants_ to look them up by name, they can always collect the `&[Wildcard]` into a `HashMap` themselves.

Reply:
I think that your original idea is probably the way to go in the long run where the embedder presents a list of "I expect these things" and then later you get a list in the same order. Otherwise if simply presented a list of what a component instance has an embedder would have to do a bunch of string comparisons to figure out which one is the one they want.

In the interim I think it's reasonable to preserve the functionality of "invoke this specific export" in some sort of mostly-ergonomic fashion, aka keeping the hash map for that.

Reply:
I'm just thinking my original idea only moves the string lookup elsewhere -- the code that translates a `&[&str]` to a `Vec<WildcardMatch>` would itself need to do a lookup for each name in order to populate the result.

Reply:
Oh so string lookups at `WildcardMatch` time I think is fine, it's just that after you have an `InstancePre<T>` you should theoretically absolutely minimize string lookups since that's the thing which is repeatedly instantiated and is perf-sensitive.

So yeah what you proposed I think would work, adding `fn name() -> &str` and then returning the same-ordered list. To reiterate though it's not necessarily critical to fix this now as I'm sure there's other more pressing things to take care of.